src/GafferOSL/ShadingEngine.cpp

```diff
@@ -289,12 +289,12 @@ class RenderState
 				userData.dataView = IECoreImage::OpenImageIOAlgo::DataView( it->second.get(), /* createUStrings = */ true );
 				if( userData.dataView.data )
 				{
+					userData.numValues = std::max( userData.dataView.type.arraylen, 1 );
```

Comment:
Silly Q: do we need to handle the case of an empty array here?

Reply:
That is a very good question, and it took me a while to figure out the answer. OIIO uses `arraylen == 0` to mean "not an array", so the short answer is the code is OK. But DataView isn't handling zero-length vectors correctly, because it reports them with `arraylen == 0` and with inconsistent values for `data` :

- For a default-constructed empty vector it reports `data == nullptr`, so the above code works.
- For a vector that has been allocated but resized to 0 (so `capacity() != 0`), it will report `data != nullptr`, so the above code will read one item of junk (but not crash).

I think we should definitely fix DataView to consistently report `data == nullptr` because there is no valid data to read from a zero-length array : it may not even have been allocated. So the short-ish answer is still that the code is OK.

But OIIO also allows "unsized" arrays by specifying `arraylen == -1`, so you could argue that we should report that in addition to `data == nullptr`. If we did that, the code above still works just fine, but our treatment of `arraylen` is a little misleading because it ignores the `arraylen == -1` case and relies on the `data != nullptr` check to guard it. So the longer answer is still that the code is OK, but it could do with a comment to clarify things (which we can't really add until we fix DataView).